src/XrdOfs/XrdOfs.cc

```diff
@@ -2252,10 +2278,29 @@ int XrdOfs::rename(const char             *old_name,  // In
        evsObject->Notify(XrdOfsEvs::Mv, evInfo);
       }
 
+// If we cannot overwrite, we must open-exclusive first.  This will test whether
+// we will destroy data in the rename (without actually destroying data).
+//
+   std::unique_ptr<XrdSfsFile> tmp_fp = nullptr;
+   if (cannot_overwrite)
+      {tmp_fp.reset(newFile(einfo));
+       if (!tmp_fp)
+          {return fsError(einfo, ENOMEM);}
+       if (SFS_OK != tmp_fp->open(new_name, SFS_O_CREAT, 0700, client, infoN))
```

Comment:
Yup - I was basically trying to balance between "introduces race condition" and "potentially leaves files dangling" (in addition to the crash, consider the fact that if the rename fails that the client may not be authorized to delete the empty file it just created).

I chose "potentially leaves files dangling" because that approach guarantees no data is destroyed, even in weird race cases.

I don't feel particularly strongly here and would be OK with switching to an existence check if you prefer that direction... just wanted to point out the downsides of the alternate approach.

Reply:
Latest commit changes to the suggested test-then-rename approach.  Resolving conversation.

Reply:
Well, all of the side-effects of surreptitiously creating a file during rename certainly makes it unappealing.  I'd really prefer an existence check. That way we get a reasonably predictable outcome regardless of what happens to the server. Frankly, the race condition hasn't been a problem in 20 years (yeah, famous last words).